backend/metasfresh-webui-api/src/main/java/de/metas/ui/web/comments/CommentsService.java

```diff
@@ -22,37 +22,104 @@
 import de.metas.comments.CommentEntry;
 import de.metas.comments.CommentEntryRepository;
 import de.metas.ui.web.comments.json.JSONComment;
 import de.metas.ui.web.comments.json.JSONCommentCreateRequest;
+import de.metas.ui.web.view.IViewRow;
+import de.metas.ui.web.window.datatypes.DocumentPath;
 import de.metas.ui.web.window.datatypes.json.DateTimeConverters;
+import de.metas.ui.web.window.descriptor.factory.DocumentDescriptorFactory;
 import de.metas.user.api.IUserDAO;
 import de.metas.util.GuavaCollectors;
+import de.metas.util.ImmutableMapEntry;
 import de.metas.util.Services;
 import lombok.NonNull;
 import org.adempiere.util.lang.impl.TableRecordReference;
 import org.springframework.stereotype.Service;
 
+import javax.annotation.Nullable;
 import java.time.ZoneId;
+import java.util.Collection;
+import java.util.Collections;
 import java.util.Comparator;
+import java.util.IdentityHashMap;
 import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Stream;
 
 @Service
 public class CommentsService
 {
 	private final CommentEntryRepository commentEntryRepository;
+	private final DocumentDescriptorFactory documentDescriptorFactory;
 	final IUserDAO userDAO = Services.get(IUserDAO.class);
 
-	public CommentsService(final CommentEntryRepository commentEntryRepository)
+	public static final IdentityHashMap<IViewRow, Boolean> NO_COMMENTS = new IdentityHashMap<>(0);
+
+	public CommentsService(final CommentEntryRepository commentEntryRepository, final DocumentDescriptorFactory documentDescriptorFactory)
 	{
 		this.commentEntryRepository = commentEntryRepository;
+		this.documentDescriptorFactory = documentDescriptorFactory;
+	}
+
+	@NonNull
+	public final IdentityHashMap<IViewRow, Boolean> hasComments(@Nullable final IViewRow row)
+	{
+		if (row == null)
+		{
+			return NO_COMMENTS;
+		}
+
+		return hasComments(Collections.singletonList(row));
+	}
+
+	@NonNull
+	public final IdentityHashMap<IViewRow, Boolean> hasComments(@NonNull final Collection<? extends IViewRow> rows)
+	{
+		if (rows.isEmpty())
+		{
+			return NO_COMMENTS;
+		}
+
+		final ImmutableMap<IViewRow, TableRecordReference> rowsForReferences = rows.stream()
+				.flatMap(IViewRow::streamRecursive)
+				.flatMap(this::toStreamOfValidTableReferences)
+				.collect(GuavaCollectors.toImmutableMap());
+
+		final Map<TableRecordReference, Boolean> referencesWithComments = commentEntryRepository.hasComments(rowsForReferences.values());
+
+		final IdentityHashMap<IViewRow, Boolean> result = new IdentityHashMap<>();
+		for (final IViewRow row : rows)
+		{
+			final TableRecordReference ref = rowsForReferences.get(row);
+			result.put(row, referencesWithComments.getOrDefault(ref, false));
+		}
+
+		return result;
 	}
 
 	@NonNull
-	public List<JSONComment> getCommentsFor(@NonNull final TableRecordReference tableRecordReference, final ZoneId zoneId)
+	public Boolean hasComments(@NonNull final DocumentPath documentPath)
```

Comment:
why `Boolean` object and not primitive `boolean`?

Reply:
mistake